<a href="https://colab.research.google.com/github/lkostenko/python_da_homeworks/blob/main/hw_3_2_%D0%9E%D1%81%D0%BD%D0%BE%D0%B2%D0%B8_%D1%80%D0%BE%D0%B1%D0%BE%D1%82%D0%B8_%D0%B7_%D0%B4%D0%B0%D0%BD%D0%B8%D0%BC%D0%B8_%D1%83_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
data_path = '/content/drive/MyDrive/курс Аналітик/навчання/Модуль 3. Робота з табличними даними: Pandas та Numpy/data/supermarket_sales.csv'
df = pd.read_csv(data_path)

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [ ]:
df[0:4]

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4

In [ ]:
df.head(4)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4

In [ ]:
df.tail(4)

Invoice ID Branch      City Customer type  Gender         Product line  \
996  303-96-2227      B  Mandalay        Normal  Female   Home and lifestyle   
997  727-02-1313      A    Yangon        Member    Male   Food and beverages   
998  347-56-2442      A    Yangon        Normal    Male   Home and lifestyle   
999  849-09-3807      A    Yangon        Member  Female  Fashion accessories   

     Unit price  Quantity  Tax 5%     Total       Date   Time  Payment  \
996       97.38        10  48.690  1022.490   3/2/2019  17:16  Ewallet   
997       31.84         1   1.592    33.432   2/9/2019  13:22     Cash   
998       65.82         1   3.291    69.111  2/22/2019  15:33     Cash   
999       88.34         7  30.919   649.299  2/18/2019  13:28     Cash   

       cogs  gross margin percentage  gross income  Rating  
996  973.80                 4.761905        48.690     4.4  
997   31.84                 4.761905         1.592     7.7  
998   65.82                 4.761905         3.291     4.1  
999  618.38                 4.761905        30.919     6.6

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [ ]:
df[615:620]

Invoice ID Branch       City Customer type  Gender  \
615  509-29-3912      A     Yangon        Member  Female   
616  327-40-9673      B   Mandalay        Member    Male   
617  840-19-2096      C  Naypyitaw        Member    Male   
618  828-46-6863      A     Yangon        Member    Male   
619  641-96-3695      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
615       Sports and travel       38.72         9  17.4240  365.9040   
616       Sports and travel       72.60         6  21.7800  457.3800   
617  Electronic accessories       87.91         5  21.9775  461.5275   
618      Food and beverages       98.53         6  29.5590  620.7390   
619     Fashion accessories       43.46         6  13.0380  273.7980   

          Date   Time      Payment    cogs  gross margin percentage  \
615  3/20/2019  12:24      Ewallet  348.48                 4.761905   
616  1/13/2019  19:51         Cash  435.60                 4.761905   
617  3/14/2019  18:10      Ewallet  439.55                 4.761905   
618  1/23/2019  11:22  Credit card  591.18                 4.761905   
619   2/7/2019  17:55      Ewallet  260.76                 4.761905   

     gross income  Rating  
615       17.4240     4.2  
616       21.7800     6.9  
617       21.9775     4.4  
618       29.5590     4.0  
619       13.0380     8.5

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1Otufe_TXRmjZxWigzBiVBLCAVUfEReXM)

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [ ]:
df.dtypes

Invoice ID                  object
Branch                      object
City                        object
Customer type               object
Gender                      object
Product line                object
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                        object
Time                        object
Payment                     object
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object

Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [ ]:
df['gross income'].describe(percentiles=[.01, .25, .5, .75, .99])

count    1000.000000
mean       15.379369
std        11.708825
min         0.508500
1%          0.913735
25%         5.924875
50%        12.088000
75%        22.445250
99%        45.250750
max        49.650000
Name: gross income, dtype: float64

75отму персентилю колонки gross income відповідає - 22.445250. Це означає, що у 75% людей дохід менше, ніж або рівний 22.45


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1bDa1YFUWDsPm78Lnpy2geQND8M1HEQiT)

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [4]:
df.City.value_counts()

City
Yangon       340
Mandalay     332
Naypyitaw    328
Name: count, dtype: int64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1G8c6M875NOW7oU3OtZegTOnyD2y8lhAV)

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [5]:
city_count = df.City.value_counts()
city_count / city_count.sum()

City
Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: count, dtype: float64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1yZZpWhL2y8DX9rSRioYCRh1VbJgcf3fY)

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [13]:
mean_yangon = df[df.City =='Yangon'] ["gross income"].describe()["mean"]
mean_mandalay = df[df.City =='Mandalay'] ["gross income"].describe()["mean"]
print(f"Cереднє значення gross income для Yangon = {mean_yangon}, cереднє значення gross income для Mandalay = {mean_mandalay}.")

if mean_yangon >  mean_mandalay:
  print(f"Cереднє значення gross income більше в місті Yangon")
else:
  print(f"Cереднє значення gross income більше в місті Mandalay")


Cереднє значення gross income для Yangon = 14.874001470588235, cереднє значення gross income для Mandalay = 15.232024096385542.
Cереднє значення gross income більше в місті Mandalay


1. **Наявність широкого асортимент товару та свіжість продукції**

Якщо магазин пропонує широкий асортимент товарів, популярні фірми / бренди, імпортні товари - це може збільшити кількість покупців. Покупці зазвичай зацікавлені в свіжих продуктах.


*Наявні дані:* за цими даними ми зможемо лише аналізувати середній gross income для кожної товарної лінії (Product line). Це покаже, чи магазини з більшим асортиментом товарів можуть мати більший  gross income.

*Додаткові:* оцінити терміни зберігання і продажу свіжих продуктів у різних магазинах.  

*Висновок:* якщо деякі товари мають вищі середні значення gross income, це може підтвердити, що асортимент товарів впливає на грос дохід магазину.
Якщо свіжіть продуктів дійсно впливає на дохід, то, можливо, потрібно зосередиться на якості пордуктів, а не на кількості. Тобто замовляти товари меншими партіями, але частіше.

2. **Локація та ремонт магазину**

Магазини, які розташовані в густонаселених або багатших райнох, можуть мати більше покупців. Також на кількість покупців може впливати інтер'єр. Люди зазвичай проводять більше часу в гарних / нових магазинах.

*Наявні дані:*
За цими даними ми можемо аналізувати середнє значення gross income для міст. Якщо магазини в певних містах мають вищі gross income, це може бути пов'язано з відмінностями в попиті та доступності клієнтів.

*Додаткові дані:*
Порівняти доходи до і після ремонту / зміни інтер'єру/ перестановки місцями певних категорій продуктів в різних магазинах.

*Висновок:* Якщо виявиться, що певне місто має значно вищий середній валовий дохід, це може підтвердити, що локація впливає на прибутковість.Можна орієнтуватися на відкриття нових магазинів у найбільш відвідуваних чи вигідних районах

3. **Акції та реклама**

Ефективна реклама може значно підвищити впізнаваність магазину та збільшити потік клієнтів, що веде до росту доходів. Проведення акцій і знижок може збільшити попит клієнтів

*Наявні дані:*
У нас немає даних про акції та знижки.

*Додаткові:*
Потрібно взяти дані за певний період в магазинах, де проводились знижки та була реклама та проаналізувати, як це вплинуло на gross income.

*Висновок:*
Якщо акції справді збільшують доходи,то можна проводити більше акцій або спеціальних пропозицій для залучення клієнтів. Якщо реклама має значний вплив, можна збільшити бюджет на рекламу


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1JdXUmK46WcoXUJYP7NS2SJlr2XjBbpQ8)

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [17]:
df["gross margin percentage"].unique()

array([4.76190476])

в стовпці gross margin percentage всі значення однакові й рівні 4.76190476

Очікуваний результат: відповідь щодо маржі в усіх інвойсах на основі отриманих результатів

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [16]:
product_line_yangon_count = df[df.City =='Yangon']["Product line"].value_counts()
percentage_prod_line_yangon = product_line_yangon_count / product_line_yangon_count.sum()
product_line_mandalay_count = df[df.City =='Mandalay']["Product line"].value_counts()
percentage_prod_line_mandalay = product_line_mandalay_count / product_line_mandalay_count.sum()
print(f"Відсотковий розпоіділ кількості інвойсів по категоріям товарів Product line в місті Yangon: {percentage_prod_line_yangon}")
print(f"Відсотковий розпоіділ кількості інвойсів по категоріям товарів Product line в місті Madalay: {percentage_prod_line_mandalay}")

Відсотковий розпоіділ кількості інвойсів по категоріям товарів Product line в місті Yangon: Product line
Home and lifestyle        0.191176
Electronic accessories    0.176471
Sports and travel         0.173529
Food and beverages        0.170588
Fashion accessories       0.150000
Health and beauty         0.138235
Name: count, dtype: float64
Відсотковий розпоіділ кількості інвойсів по категоріям товарів Product line в місті Madalay: Product line
Fashion accessories       0.186747
Sports and travel         0.186747
Electronic accessories    0.165663
Health and beauty         0.159639
Food and beverages        0.150602
Home and lifestyle        0.150602
Name: count, dtype: float64


Очікуваний результат для кожного з міст:

![](https://drive.google.com/uc?export=view&id=1IbG58YK5KVqC-wCnllF1btgEgfWTbi_O)

![](https://drive.google.com/uc?export=view&id=1aKOkLmlMq1sOh4GbwyaQyjfhNJzdU4Yh)